firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -159,64 +161,97 @@ public static FirebaseInstallations getInstance(@NonNull FirebaseApp app) {
-  @VisibleForTesting
-  String getName() {
-    return firebaseApp.getName();
+  private Task<InstallationTokenResult> addGetAuthTokenListener(
+      @AuthTokenOption int authTokenOption) {
+    TaskCompletionSource<InstallationTokenResult> taskCompletionSource =
+        new TaskCompletionSource<>();
+    StateListener l = new GetAuthTokenListener(utils, taskCompletionSource);
+    synchronized (lock) {
+      if (authTokenOption == FORCE_REFRESH) {
+        shouldRefreshAuthToken = true;
+      }
+      listeners.add(l);
+    }
+    return taskCompletionSource.getTask();
   }
 
-  /**
-   * Returns the {@link PersistedFidEntry} from shared prefs.
-   *
-   * @throws {@link FirebaseInstallationsException} when shared pref is empty or {@link
-   *     PersistedFidEntry} is in error state.
-   */
-  private PersistedFidEntry getPersistedFid() throws FirebaseInstallationsException {
-    PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
-    if (persistedFidMissingOrInErrorState(persistedFidEntry)) {
-      throw new FirebaseInstallationsException(
-          "Failed to get existing fid.", FirebaseInstallationsException.Status.CLIENT_ERROR);
+  private void triggerOnStateReached(PersistedFidEntry persistedFidEntry) {
+    synchronized (lock) {
+      Iterator<StateListener> it = listeners.iterator();
+      while (it.hasNext()) {
+        StateListener l = it.next();
+        boolean doneListening = l.onStateReached(persistedFidEntry);
+        if (doneListening) {
+          it.remove();
+        }
+      }
     }
-    return persistedFidEntry;
   }
 
-  private static boolean persistedFidMissingOrInErrorState(PersistedFidEntry persistedFidEntry) {
-    return persistedFidEntry == null
-        || persistedFidEntry.getRegistrationStatus() == RegistrationStatus.REGISTER_ERROR;
-  }
+  private final void doRegistration() {
+    try {
+      PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
 
-  @NonNull
-  private static <F, T> Continuation<F, T> orElse(@NonNull Supplier<T> supplier) {
-    return t -> {
-      if (t.isSuccessful()) {
-        return (T) t.getResult();
+      // New FID needs to be created
+      if (persistedFidEntry.isErrored() || persistedFidEntry.isNotGenerated()) {
+        String fid = utils.createRandomFid();
+        persistFid(fid);
+        persistedFidEntry = persistedFid.readPersistedFidEntryValue();
       }
-      return supplier.get();
-    };
-  }
 
-  @NonNull
-  private static <F, T> Continuation<F, T> awaitFidRegistration(
-      @NonNull Supplier<T> supplier, AwaitListener listener) {
-    return t -> {
-      // Waiting for Task that registers FID on the FIS Servers
-      listener.await(AWAIT_TIMEOUT_IN_SECS, TimeUnit.SECONDS);
-      return supplier.get();
-    };
-  }
+      // Always notify the GetIdListeners. For GetAuthTokenListeners, only notify if force
```

Comment:
I am concerned that now a getId call won't complete as early as it could if there's a pending force-refresh call. Is that something we should be worried about?

Reply:
We need not be worries. getId call will never be called with force-refresh.

Reply:
Thanks for clarifying offline. I understand your concern now. I referred to your demo, made some modifications and updated the code. PTAL and let me wdyt.

Thank you.